File: 890-find-and-replace-pattern/890-find-and-replace-pattern.cpp
class Solution {
public:
    
    bool isMatching(string word, string pattern){
        
        map<char, char> mp;
        set<char> s;
        
        for(int i=0; i<word.size(); i++){
            
            if(mp.find(pattern[i]) != mp.end()){
                if(mp[pattern[i]] != word[i]){
                    return false;
                }
            }
            else{
                if(s.find(word[i]) != s.end()) return false;
                
                mp[pattern[i]] = word[i];
                s.insert(word[i]);
            }
        }
        
        return true;
    }
    vector<string> findAndReplacePattern(vector<string>& words, string pattern) {
        
        vector<string> res;
        
        for(auto word : words){
            
            if(isMatching(word, pattern) == true){
                res.push_back(word);
            }
        }
        
        return res;
    }
};

File: 890-find-and-replace-pattern/890-find-and-replace-pattern.py
class Solution:
    def isMatching(self, word, pattern):
        
        patternMapping = {}
        wordSet = set()
        
        for i in range(len(pattern)):
            wch = word[i]
            pch = pattern[i]
            
            if pch in patternMapping:
                if patternMapping[pch] != wch:
                    return False
            else:
                if wch in wordSet:
                    return False
                patternMapping[pch] = wch
                wordSet.add(wch)
        return True
    
    def findAndReplacePattern(self, words: List[str], pattern: str) -> List[str]:
        
        res = []
        for word in words:
            if self.isMatching(word, pattern):
                res.append(word)
        
        return res
        